### PR TITLE
Alec - Added PUT (edit) functionality to Students

### LINKED ISSUE #28: SUBFEATURE: Add PUT (edit) endpoint for a single record in Students table
# Dependencies

This issue should be completed *after* #26 

# Overview

In this issue, we add one more method from the example Controller used as a basis for our code; namely the method with the annotation shown below:

| Annotation | Purpose |
|-|-|
|`@PutMapping("")` | Get a single record from the table; use the value passed in as a `@RequestParam` to do a lookup by id.  If a matching row is found, update that row with the values passed in as a JSON object. If a matching row is not found, throw an `EntityNotFoundException`. |
           

Note that while it isn't strictly necessary, it may be a good idea to implement `GET` for a single item
first; when using `PUT` manually, it's convenient to be able to copy/paste the JSON representation of the data from the
`GET` by id method, paste it into the swagger interface, then edit it to make the changes you want before clicking to
submit the request.

You'll need to copy over the code for the PUT method from the example controller, 
and then paste those into `HelpRequestController.java` and edit them to match your database table.

Then, copy over the tests that pertain to the PUT method from the example controller,
paste those into `HelpRequestControllerTests.java` and edit them to match your database table.

# Acceptance Criteria:

- [x] In `HelpRequestController.java` there is code for an 
      endpoint `PUT /api/HelpRequest?id=123` endpoint 
      that accepts JSON for a new set of values for the database
      fields other than `id`, and updates the values of those fields.
- [ ] The Swagger-UI endpoints for this endpoint is well documented
      so that any member of the team can understand how to use it.
- [ ] The endpoint works as expected on localhost.
- [ ] The endpoint works as expected when deployed to Dokku.
- [x] There is full test coverage (Jacoco) for the new code in 
      `HelpRequestController.java`
- [ ] There is full mutation test coverage (Pitest) for new code in
      `HelpRequestController.java`   

# What to do next

Do a PR, following all of the usual steps (title, description, `Closes #n` text,
dragging issue to "In Review", requesting reviewers).

Check to see if any fellow team members PRs need to be reviewed.

Then, assign yourself the next incomplete issue for your database table (`HelpRequest`)

Follow all the usual steps (assign to self, drag to "In Progress", start new branch.)

When you are done with all issues for `HelpRequest`, please give your full attention to 
helping others on your team to complete the sprint, so that the team
is ready to submit on Canvas by the deadline.

### PR DESCRIPTION
**Pull request by alecsong222**
## Overview

This PR adds the Edit functionality to Students.
Branch - https://github.com/ucsb-cs156-f24/proj-happycows-f24-09/tree/Alec-StudentsEdit

## Screenshots (Optional)

image

## Validation (Optional)

Navigate to https://happycows-dev-alecsong222-students.dokku-09.cs.ucsb.edu/
Create a student
Navigate to PUT and Execute without filling in the id; ensure the following error shows up
image
Input your created student's ID, make edits, and ensure that the student has been edited
Use List all Students or Get a single Student by ID to ensure that your edits have been confirmed

## Tests

- [x] Backend Unit tests (`mvn test`) pass
- [x] Backend Test coverage (`mvn test jacoco:report`) 100%
- [x] Backend Mutation tests (`mvn test pitest:mutationCoverage`) 100% 
- [x] Frontend Unit tests (`npm test`) pass
- [x] Frontend Test coverage (`npm run coverage`) 100%
- [x] Frontend Mutation tests (`npx stryker run`) 100% 
- [x] Frontend Linting (`npx eslint --fix src`) 

## Linked Issues

Closes #28 

## PR Checklist
- [x] Commented code removed
- [x] PR deals with only one concern
- [x] PR Title is descriptive enough
- [x] PR description details what a user/dev will notice is different
- [x] No merge conflicts
- [x] Deployed and linked with Dokku deployment
- [x] New branch updated with main
- [x] PR is linked to a branch
- [x] PR is linked to an issue/ticket
- [x] Linked issue is in "In Review" column on Kanban board
- [x] PR is tagged with the Team's tag
- [x] All test cases pass

### CODE REVIEW

**Review by TomShangguan — DISMISSED** [review]
LGTM! I looked and tried in dokku. It works.

File: frontend/src/main/components/Leaderboard/LeaderboardTable.js [hunk 1]
```diff
@@ -18,12 +18,9 @@ export default function LeaderboardTable({ leaderboardUsers }) {
...
-            accessor: (row, _rowIndex) => {
-                return USD.format(row.totalWealth);
-            },
+            accessor: "totalWealth",
             Cell: (props) => {
-                return <div style={{ textAlign: "right" }}>{props.value}</div>;
+                return <div style={{ textAlign: "right" }}>{(USD.format(props.value))}</div>;
             },
         },
         {
```
Comment (pconrad) [inline frontend/src/main/components/Leaderboard/LeaderboardTable.js]: What is this change doing in this PR?  

It should not be in this PR.

File: src/main/java/edu/ucsb/cs156/happiercows/controllers/StudentsController.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    @Autowired
+    StudentsRepository StudentsRepository;
+
+    /**
+     * List all students
+     * 
+     * @return an iterable of Students
+     */
+    @Operation(summary= "List all students")
+    @PreAuthorize("hasRole('ROLE_USER')")
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/happiercows/controllers/StudentsController.java]:
```suggestion
    @PreAuthorize("hasRole('ROLE_ADMIN')")
```

We should not be releasing lists of students to anyone other than the Admins.

File: src/main/java/edu/ucsb/cs156/happiercows/controllers/StudentsController.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    /**
+     * Get a student by id
+     * 
+     * @param id the id of the student
+     * @return a Student
+     */
+    @Operation(summary = "Get a single Student by id")
+    @PreAuthorize("hasRole('ROLE_USER')")
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/happiercows/controllers/StudentsController.java]:
```suggestion
    @PreAuthorize("hasRole('ROLE_ADMIN')")
```

Only admins need this ability; they are the ones managing course rosters.   There may be individual backend functions available to users where the Students and Course tables will be consulted, but that's different; here we are doing straight lookups, which should be ADMIN only.

File: src/main/java/edu/ucsb/cs156/happiercows/controllers/StudentsController.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+            @Parameter(name="id") @RequestParam Long id,
+            @RequestBody @Valid Students incoming) {
+
+        Students student = StudentsRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException(Students.class, id));
+
+        student.setLName(incoming.getLName());
+        student.setFmName(incoming.getFmName());
+        student.setEmail(incoming.getEmail());
+        student.setCourseId(incoming.getCourseId());
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/happiercows/controllers/StudentsController.java]: We shouldn't mess with the course id.  Once a student is added to a course, we should not change the course id.  If they were added to the wrong course they should be deleted from the one course and added to another one.

The reasons for that are that later functionality may depend on a student object being "immutable" from the standpoint of which course they "belong to" (including, for example, adding an "Instructor" role in between ADMIN and USER that can only work with the Courses and Commons that they themselves created.

File: src/main/java/edu/ucsb/cs156/happiercows/entities/Students.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+public class Students {
+
+  // Unique Student Id
+  @Id
+  @GeneratedValue(strategy = GenerationType.IDENTITY)
+  private long id;
+
+  private String perm;
+  private String lName;
+  private String fmName;
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/happiercows/entities/Students.java]: See Code Reviews on PR #37

File: src/main/resources/db/migration/changes/Students.json [hunk 6]
```diff
@@ -0,0 +1,74 @@
+{
+    "databaseChangeLog": [
+      {
+        "changeSet": {
+          "id": "1234",
+          "author": "Alec Song",
```
Comment (pconrad) [inline src/main/resources/db/migration/changes/Students.json]: We typically use the github id here.

File: src/main/resources/db/migration/changes/Students.json [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+                      "name": "LNAME",
+                      "type": "VARCHAR(255)"
+                    }
+                  },
+                  {
+                    "column": {
+                      "name": "FMNAME",
+                      "type": "VARCHAR(255)"
+                    }
+                  },
```
Comment (pconrad) [inline src/main/resources/db/migration/changes/Students.json]: See Code Reviews on PR #37

File: src/test/java/edu/ucsb/cs156/happiercows/controllers/StudentsControllerTests.java [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+                mockMvc.perform(get("/api/Students/all"))
+                                .andExpect(status().is(403)); // logged out users can't get all
+        }
+
+        @WithMockUser(roles = { "USER" })
+        @Test
+        public void logged_in_users_can_get_all() throws Exception {
+                mockMvc.perform(get("/api/Students/all"))
+                                .andExpect(status().is(200)); // logged
+        }
```
Comment (pconrad) [inline src/test/java/edu/ucsb/cs156/happiercows/controllers/StudentsControllerTests.java]: We would actually want ADMIN to get a 200, and USER to get a 403

File: src/test/java/edu/ucsb/cs156/happiercows/controllers/StudentsControllerTests.java [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+
+        @WithMockUser(roles = { "USER" })
+        @Test
+        public void logged_in_regular_users_cannot_post() throws Exception {
+                mockMvc.perform(post("/api/Students/post"))
+                                .andExpect(status().is(403)); // only admins can post
+        }
+
+        @WithMockUser(roles = { "USER" })
```
Comment (pconrad) [inline src/test/java/edu/ucsb/cs156/happiercows/controllers/StudentsControllerTests.java]:
```suggestion
        @WithMockUser(roles = { "ADMIN" })
```

File: src/test/java/edu/ucsb/cs156/happiercows/controllers/StudentsControllerTests.java [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+
+                // assert
+
+                verify(StudentsRepository, times(1)).findById(eq(7L));
+                String expectedJson = mapper.writeValueAsString(student);
+                String responseString = response.getResponse().getContentAsString();
+                assertEquals(expectedJson, responseString);
+        }
+
+        @WithMockUser(roles = { "USER" })
```
Comment (pconrad) [inline src/test/java/edu/ucsb/cs156/happiercows/controllers/StudentsControllerTests.java]:
```suggestion
        @WithMockUser(roles = { "ADMIN" })
```

File: src/test/java/edu/ucsb/cs156/happiercows/controllers/StudentsControllerTests.java [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+
+                verify(StudentsRepository, times(1)).findById(eq(7L));
+                Map<String, Object> json = responseToJson(response);
+                assertEquals("EntityNotFoundException", json.get("type"));
+                assertEquals("Students with id 7 not found", json.get("message"));
+        }
+
+
+        
+        @WithMockUser(roles = { "USER" })
```
Comment (pconrad) [inline src/test/java/edu/ucsb/cs156/happiercows/controllers/StudentsControllerTests.java]:
```suggestion
        @WithMockUser(roles = { "ADMIN" })
```

File: src/test/java/edu/ucsb/cs156/happiercows/controllers/StudentsControllerTests.java [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+        @Test
+        public void logged_in_user_can_get_all_Students() throws Exception {
```
Comment (pconrad) [inline src/test/java/edu/ucsb/cs156/happiercows/controllers/StudentsControllerTests.java]:
```suggestion
        public void logged_in_admin_can_get_all_Students() throws Exception {
```

**Review by pconrad — CHANGES_REQUESTED** [review]
A few permission changes, and the unrelated code from PR #25

**Review by alecsong222 — COMMENTED** [review]
implemented all suggested changes

**Review by pconrad — APPROVED** [review]
LGTM

**Pull request merged**